[RPC Framework] Separate initialize_from_module_rref method out of RemoteModule constructor

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59292 [RPC Framework] Separate initialize_from_module_rref method out of RemoteModule constructor**

#Closes: https://github.com/pytorch/pytorch/issues/58274

Create an alternate initialization method, and also create a few util functions to avoid duplicate code.

Differential Revision: [D28825895](https://our.internmc.facebook.com/intern/diff/D28825895/)

Commit:
[RPC Framework] Separate initialize_from_module_rref method out of RemoteModule constructor

#Closes: https://github.com/pytorch/pytorch/issues/58274

Create an alternate initialization method, and also create a few util functions to avoid duplicate code.

Differential Revision: [D28825895](https://our.internmc.facebook.com/intern/diff/D28825895/)

[ghstack-poisoned]

Commit:
Update on "[RPC Framework] Separate initialize_from_module_rref method out of RemoteModule constructor"


#Closes: https://github.com/pytorch/pytorch/issues/58274

Create an alternate initialization method, and also create a few util functions to avoid duplicate code.

Differential Revision: [D28825895](https://our.internmc.facebook.com/intern/diff/D28825895/)

[ghstack-poisoned]

Commit:
Update on "[RPC Framework] Separate initialize_from_module_rref method out of RemoteModule constructor"


#Closes: https://github.com/pytorch/pytorch/issues/58274

Create an alternate initialization method, and also create a few util functions to avoid duplicate code.

Differential Revision: [D28825895](https://our.internmc.facebook.com/intern/diff/D28825895/)

[ghstack-poisoned]

wanchaol commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -214,105 +209,155 @@ def __init__(
+            >>> )
+            >>>
+            >>> def create_remote_module_by_module_rref(remote_device, module_rref):
+            >>>     remote_module = object.__new__(RemoteModule)
```
can we make this `__new__` inside the `init_from_module_rref` instead? i.e. making this function as static function and generate a RemoteModule internally, then return it? A single API should be more intuitive to the user, the UX will be like:

```
remote_module = RemoteModule.init_from_module_rref(remote_device, module_rref)
```

SciPioneer replied:
Sounds good! Actually this change can also save the work of creating a separate RPC handler.

wanchaol commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -214,105 +209,155 @@ def __init__(
+                In addition, the device field can be optional and the default value is "cpu".
+            module_rref (RRef[nn.Module]): The module reference shared by both the caller and
+                the created remote module.
+            _module_interface_cls (type, optional): The TorchScript interface type for the module
```
Let's not document this arg as we might change the way to initialize a scriptModule based RemoteModule(ultimately we would like to remove this and make sure no discrepency to eager mode API).

SciPioneer replied:
Similar to the comments in the constructor, we have one line here:
"Warning, this is an experimental API and susceptible to frequent changes."

wanchaol reviewed (commented):
Looks great! Thanks for working on this, only have some suggestions on the API.

wanchaol commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -214,105 +209,155 @@ def __init__(
-                )
+    def init_from_module_rref(
+        self,
+        remote_device: str,
```
Is this `remote_device` a necessary argument? from the example I see that we are duplicating the "worker1/cpu". I think `init_from_module_rref` is a API that tries to initialize RemoteModule locally on the current worker, and since we actually don't the real RemoteModule (the real one is the RRef[Module] passed in), so it seems we don't need to worry about the device that we need to place into?

SciPioneer replied:
It's needed to infer `enable_moving_cpu_tensors_to_cuda` and hence affect `_init_template`.

wanchaol replied:
I see, I guess it's inevictable as RRef itself doesn't store information like this.

Commit:
Update on "[RPC Framework] Separate initialize_from_module_rref method out of RemoteModule constructor"


#Closes: https://github.com/pytorch/pytorch/issues/58274

Create an alternate initialization method, and also create a few util functions to avoid duplicate code.

Differential Revision: [D28825895](https://our.internmc.facebook.com/intern/diff/D28825895/)

[ghstack-poisoned]

wanchaol reviewed (approved):
lgtm, thanks!
